core/peerset/src/slots.rs

```diff
@@ -0,0 +1,132 @@
+// Copyright 2018-2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::mem;
+use libp2p::PeerId;
+use linked_hash_map::LinkedHashMap;
+
+/// Describes the nature of connection with a given peer.
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum SlotType {
+	/// Reserved peer is a peer we should always stay connected to.
+	Reserved,
+	/// Common peer is a type of peer that we stay connected to only if it's
+	/// useful for us.
+	Common,
+}
+
+/// Descibes why the reason of not being able to add given peer.
+pub enum SlotError {
+	/// Error returned when we are already connected to this peer.
+	AlreadyConnected(PeerId),
+	/// Error returned when max number of connections has been already established.
+	MaxConnections(PeerId),
+	/// Error returned when we should disconnect from a given common peer to make space
+	/// for a reserved peer.
+	DemandReroute {
+		/// Peer we should disconnect from.
+		disconnect: PeerId,
+		/// Peer we should connect to.
+		connect: PeerId,
+	}
+}
+
+/// Contains all information about group of slots.
+#[derive(Debug)]
+pub struct Slots {
+	max_slots: usize,
+	/// Nodes and their type. We use `LinkedHashMap` to make this data structure more predictable
+	slots: LinkedHashMap<PeerId, SlotType>,
+}
+
+impl Slots {
+	/// Creates a group of slots with a limited size.
+	pub fn new(max_slots: u32) -> Self {
+		let max_slots = max_slots as usize;
+		Slots {
+			max_slots,
+			slots: LinkedHashMap::with_capacity(max_slots),
+		}
+	}
+
+	/// Returns true if one of the slots contains given peer.
+	pub fn contains(&self, peer_id: &PeerId) -> bool {
+		self.slots.contains_key(peer_id)
+	}
+
+	/// Returns Ok if we successfully connected to a given peer.
+	pub fn add_peer(&mut self, peer_id: PeerId, slot_type: SlotType) -> Result<(), SlotError>  {
```

Comment:
From looking at the return values, I'm not sure if this should be a `Result`. You return `Err` even in situations that I would consider successful and where you change the state.
That can be improved later, though.

Reply:
Fair enough, I'll change the `Result`